src/AbstractActionHandler.ts

```diff
@@ -189,22 +197,64 @@ export abstract class AbstractActionHandler {
    */
   protected async handleActions(
     state: any,
-    block: Block,
     context: any,
+    nextBlock: NextBlock,
     isReplay: boolean,
   ): Promise<void> {
+    const { block } = nextBlock
     const { blockInfo } = block
 
-    const versionedActions = await this.applyUpdaters(state, block, isReplay, context)
+    const versionedActions = await this.applyUpdaters(state, block, context, isReplay)
     if (!isReplay) {
-      this.runEffects(versionedActions, block, context)
+      this.runEffects(versionedActions, context, nextBlock)
     }
 
     await this.updateIndexState(state, block, isReplay, this.handlerVersionName, context)
     this.lastProcessedBlockNumber = blockInfo.blockNumber
     this.lastProcessedBlockHash = blockInfo.blockHash
   }
 
+  private range(start: number, end: number) {
+    return Array(end - start).fill(0).map((_, i: number) => i + start)
+  }
+
+  private runOrDeferEffect(
+    effect: Effect,
+    payload: any,
+    nextBlock: NextBlock,
+    context: any,
+  ) {
+    const { block, lastIrreversibleBlockNumber } = nextBlock
+    if (!effect.deferUntilIrreversible || block.blockInfo.blockNumber <= lastIrreversibleBlockNumber) {
+      effect.run(payload, block, context)
+    } else if (!this.deferredEffects[block.blockInfo.blockNumber]) {
+      this.deferredEffects[block.blockInfo.blockNumber] = [() => effect.run(payload, block, context)]
+    } else {
+      this.deferredEffects[block.blockInfo.blockNumber].push(() => effect.run(payload, block, context))
+    }
+  }
+
+  private runDeferredEffects(lastIrreversibleBlockNumber: number) {
+    const nextDeferredBlockNumber = this.getNextDeferredBlockNumber()
+    if (!nextDeferredBlockNumber) { return }
+    for (const blockNumber of this.range(nextDeferredBlockNumber, lastIrreversibleBlockNumber + 1)) {
+      if (this.deferredEffects[blockNumber]) {
+        for (const deferredEffect of this.deferredEffects[blockNumber]) {
```

Comment:
this would be such an awesome while loop, and remove the initial conditional ;) but I digress
```
while(this.deferredEffects[blockNumber] && this.deferredEffects[blockNumber].length > 0) {
 const effect = this.deferredEffects[blockNumber].shift()
 effect()
}
```

Reply:
This would still leave an empty array, and the `delete` would still be necessary